Update CI config

anupambhatnagar opened the pull request:
## What does this PR do?
Fixes Issue [#843](https://github.com/facebookresearch/fairscale/issues/843)

Due to the hardware upgrade to T4 instances some tests with timers/thresholds were failing/flaky. I made small changes to make sure they pass consistently. Here's the list of tests.

increased timeout to 70 seconds for these tests (see `command` in `run_unittests_from_list`):
- test_multiple_forward_checkpoint[bn-model1-auto_wrap_bn-flatten-full]

modified timer settings for these tests in `tests/nn/data_parallel/test_fsdp_overlap.py`:
- test_forward_overlap[mixed-flatten-2]
- test_forward_overlap[mixed-no_flatten-2]

changed tolerance to 5e-1 from 2e-1 for these tests in `tests/nn/pipe/skip/test_gpipe.py`:
- test_1to3[never-3]
- test_1to3[never-1:2]
- test_1to3[never-2:1]
- test_1to3[never-1:1:1]
- test_1to3[always-3]
- test_1to3[always-1:2]
- test_1to3[always-2:1]
- test_1to3[always-1:1:1]
- test_1to3[except_last-3]
- test_1to3[except_last-1:2]
- test_1to3[except_last-2:1]
- test_1to3[except_last-1:1:1]

Additional changes:
- cuda version has been set to 11.2 for all jobs. There were silent failures with pyenv when using cuda 11.4. I may revisit cuda 11.4 at a later point. 
- in some places I have added underscores to make it easier to read the pytorch version. E.g. the string 1110 can be interpreted as 1.11.0 or 11.1.0 so adding underscores can improve readability, in my opinion. 

From here on we will be supporting the following pytorch versions:
  - latest pytorch LTS release (1.8 as of this PR)
  - latest pytorch stable release (1.10)
  - a recent pytorch nightly release

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you write any new necessary tests?
  - [x] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [x] N/A


## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

min-xu-ai reviewed (approved):
LGTM! Can you update the README to state that we now that LTS, latest and nightly? I think it used to say that we test last 3 releases or something like that. Thanks!

anj-s commented on .circleci/config.yml:
```diff
@@ -64,22 +67,8 @@ setup_venv: &setup_venv
         which pip
         pip install --upgrade pip
 
-install_dep_171: &install_dep_171
```
nit: can we add that we are updating what we support in the PR description?

anupambhatnagar replied:
done.

anj-s commented on .circleci/config.yml:
```diff
@@ -272,14 +260,14 @@ jobs:
       - restore_cache:
           keys:
-            - cache-key-cpu-py37-190-{{ checksum "setup.py"}}-{{ checksum "requirements-test.txt"}}
+            - cache-key-cpu-py37-1100-{{ checksum "setup.py"}}-{{ checksum "requirements-test.txt"}}
```
how about adding hyphens to make it clear that it corresponds to the PT version?

anupambhatnagar replied:
done

anj-s commented on .circleci/config.yml:
```diff
@@ -446,13 +391,13 @@ jobs:
         default: "/dev/non_exist"
 
-    <<: *gpu_cu111
+    <<: *gpu_cu_11_2_small_multi
```
when do we decide to use small_multi vs medium_multi?

anupambhatnagar replied:
small_multi has 2 gpus and medium_multi has 4 gpus. there are certain jobs which require world_size=4. hence the use of medium_multi for some jobs. E.g. See the logs here https://app.circleci.com/pipelines/github/facebookresearch/fairscale/3740/workflows/643c4a44-4ca6-4b8f-abe2-6554c6222c18/jobs/40210

anj-s commented on .circleci/config.yml:
```diff
@@ -627,27 +572,21 @@ workflows:
+      - gpu_tests_1_8_1:
           test_list_file: tests/ci_test_list_3.txt
-      - gpu_tests_190:
+      - gpu_tests_1_10_0:
```
Yay! all these deletions should bring down our E2E test time :)

anj-s commented:
+1 on Min's comment about adding this to the README. We should open a followup issue about deleting code/tests that require conditional logic based on previous now unsupported PyTorch versions.

anupambhatnagar commented:
@anj-s @min-xu-ai I ran into some git issues when I rebased this branch onto main (after the release). Closing this PR and opening a new one so that we can do a clean commit. I have incorporated all feedback in the comments in the new PR. Sorry about this. Hope it's not too much trouble for you to review again.
